src/odf/dataset-impl/src/entities/metadata_chain_validators.rs

```diff
@@ -856,6 +856,47 @@ impl<'a> ValidateSetDataSchemaVisitor<'a> {
             vocab: None,
         })
     }
+
+    fn check_allowed_schema_migration(diff: &DataSchemaDiff<'_>) -> bool {
+        diff.items
+            .iter()
+            .all(Self::check_allowed_schema_migration_rec)
+    }
+
+    fn check_allowed_schema_migration_rec(diff: &DataSchemaDiffItem<'_>) -> bool {
+        #[expect(clippy::match_same_arms)]
+        match diff {
+            DataSchemaDiffItem::SchemaAttributesChanged => true,
+            DataSchemaDiffItem::FieldAttributesChanged => true,
+            DataSchemaDiffItem::FieldAdded { field_rhs } => {
+                if field_rhs.is_optional() {
+                    // It's OK to introduce a new optional field as this maintains compatibility
+                    // from publisher and consumer perspective
+                    true
+                } else {
+                    false
+                }
+            }
+            DataSchemaDiffItem::FieldRemoved => false,
```

Comment:
Is field removal really a breaking change? Maybe we could just ignore the old redundand field in the old records?